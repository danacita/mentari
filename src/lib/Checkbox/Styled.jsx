import styled, { css } from 'styled-components';
import Color from '../Styles/bases/Color';
import { Checkmark } from '../icons';

const CheckboxWrapper = styled.div`
  display: inline-block;
`;

const CheckboxInput = styled.input`
  position: absolute;
  opacity: 0;
  &:focus {
    & + label {
      &:before {
        border: 1px solid;
        border-color: ${Color.blue};
      }
    }
  }
  & + label {
    cursor: pointer;
    position: relative;
    display: inline-block;
    font-weight: 400;
    margin-bottom: 0;
    line-height: 24px;
    &:before {
      content: '';
      position: absolute;
      width: 20px;
      height: 20px;
      background-color: ${Color.white};
      border: 1px solid ${Color.grey191};
      border-radius: 2px;
      display: inline-block;
    }
    ${props =>
      props.disabled
        ? css`
            color: ${Color.grey191};
            cursor: not-allowed;
            &:before {
              background-color: ${Color.grey239};
              border: 1px solid ${Color.grey216};
            }
          `
        : null};
  }
  &:checked + label {
    &:before {
      background-color: ${Color.white};
      border-color: ${Color.grey191};
    }
    &:after {
      content: '';
      position: absolute;
      width: 20px;
      height: 20px;
      background-image: url('${Checkmark}');
      background-position: center;
      background-repeat: no-repeat;
      top: 0;
    }
  }
`;

export { CheckboxWrapper, CheckboxInput };
