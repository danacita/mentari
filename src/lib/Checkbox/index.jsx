/* eslint-disable jsx-a11y/label-has-for */
import PropTypes from 'prop-types';
import React, { Fragment } from 'react';
import { CheckboxWrapper, CheckboxInput } from './Styled';
import Color from '../Styles/bases/Color';
import { Margin, Padding } from '../Styles/utils';

const propTypes = {
  id: PropTypes.string,
  testid: PropTypes.string,
  name: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  disabled: PropTypes.bool,
  withRedux: PropTypes.bool,
  checked: PropTypes.bool,
  onChange: PropTypes.func,
  // input prop is passed by redux-form to keep field in sync with state
  input: PropTypes.shape({
    onBlur: PropTypes.func,
    onChange: PropTypes.func
  }),
  meta: PropTypes.shape({
    error: PropTypes.string,
    touched: PropTypes.bool
  })
};

const defaultProps = {
  id: null,
  testid: '',
  name: null,
  label: '',
  disabled: false,
  withRedux: true,
  checked: true,
  input: {},
  meta: {},
  onChange: null
};

const Checkbox = ({
  id,
  testid,
  name,
  label,
  input,
  disabled,
  meta: { touched, error },
  withRedux,
  checked,
  onChange
}) => (
  <CheckboxWrapper>
    {withRedux ? (
      <Fragment>
        <CheckboxInput
          id={id}
          data-testid={testid}
          disabled={disabled}
          type="checkbox"
          name={name}
          value={input.value}
          onBlur={input.onBlur}
          onChange={input.onChange}
        />
        <label htmlFor={id}>
          <Padding left={28}>{label}</Padding>
        </label>
        {touched && error && (
          <Margin top={8} left={28}>
            <span style={{ color: Color.red }}>{error}</span>
          </Margin>
        )}
      </Fragment>
    ) : (
      <Fragment>
        <CheckboxInput
          id={id}
          data-testid={testid}
          disabled={disabled}
          type="checkbox"
          name={name}
          checked={checked}
          onChange={onChange}
        />
        <label htmlFor={id}>
          <Padding left={28}>{label}</Padding>
        </label>
      </Fragment>
    )}
  </CheckboxWrapper>
);

Checkbox.propTypes = propTypes;
Checkbox.defaultProps = defaultProps;

export default Checkbox;
