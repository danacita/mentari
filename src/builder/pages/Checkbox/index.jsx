import React, { Component, Fragment } from 'react';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { atomOneDarkReasonable } from 'react-syntax-highlighter/dist/esm/styles/hljs';
import { ContentWithRightBar } from '../Styled';
import RightbarShared from '../../shared/Rightbar';
import DemoShared from '../../shared/Demo';
import PropsShared from '../../shared/Props';
import CodeShared from '../../shared/Code';
import { Margin, Checkbox, TextField } from '../../../lib';

class AvatarPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      checkboxDemo: false,
      checkboxId: 'Id',
      checkboxName: 'Name',
      checkboxText: 'Checkbox',
      checkboxDisable: false
    };
  }

  render() {
    const { checkboxDemo, checkboxId, checkboxText, checkboxName, checkboxDisable } = this.state;
    const codeString = `import { Checkbox } from '@danacita/mentari';\n<Checkbox id="${checkboxId}" name="${checkboxName}" label="${checkboxText}" ${
      checkboxDisable ? `disabled ` : ''
    }/>`;
    return (
      <Fragment>
        <ContentWithRightBar>
          <h1>Checkbox</h1>
          <p>A checkbox is an input which acts as a boolean.</p>
          <Margin top={24}>
            <PropsShared>
              <li>
                <span>id</span>
                <span>Apply a id to the control element</span>
              </li>
              <li>
                <span>name</span>
                <span>Set the name of Checkbox</span>
              </li>
              <li>
                <span>disabled</span>
                <span>Apply disabled state of Checkbox for unusable and unclickable</span>
              </li>
            </PropsShared>
          </Margin>
          <DemoShared>
            <h4>Demo</h4>
            <p>You can change the props via the component builder on your right</p>
            <Margin top={16}>
              <Checkbox
                id="checkboxDemo"
                name="checkboxDemo"
                label={checkboxText}
                withRedux={false}
                checked={checkboxDemo}
                onChange={() => this.setState({ checkboxDemo: !checkboxDemo })}
                disabled={checkboxDisable}
              />
            </Margin>
          </DemoShared>
          <h4>Code</h4>
          <Margin top={16}>
            <CodeShared copyCode={codeString}>
              <SyntaxHighlighter language="jsx" style={atomOneDarkReasonable} showLineNumbers>
                {codeString}
              </SyntaxHighlighter>
            </CodeShared>
          </Margin>
          <RightbarShared title="Checkbox">
            <Margin bottom={24}>
              <TextField
                autoFocus
                id="checkboxId"
                label="Checkbox Id"
                withRedux={false}
                value={checkboxId}
                onChange={e => this.setState({ checkboxId: e.target.value })}
              />
            </Margin>
            <Margin bottom={24}>
              <TextField
                id="checkboxName"
                label="Checkbox Name"
                withRedux={false}
                value={checkboxName}
                onChange={e => this.setState({ checkboxName: e.target.value })}
              />
            </Margin>
            <Margin bottom={24}>
              <TextField
                id="checkboxText"
                label="Checkbox Text"
                withRedux={false}
                value={checkboxText}
                onChange={e => this.setState({ checkboxText: e.target.value })}
              />
            </Margin>
            <Margin bottom={16}>
              <h4>State of Checkbox</h4>
            </Margin>
            <Margin bottom={24}>
              <Checkbox
                id="buttonDisable"
                name="buttonDisable"
                label="Disable"
                checked={checkboxDisable}
                withRedux={false}
                onChange={e => this.setState({ checkboxDisable: e.target.checked })}
              />
            </Margin>
          </RightbarShared>
        </ContentWithRightBar>
      </Fragment>
    );
  }
}

export default AvatarPage;
