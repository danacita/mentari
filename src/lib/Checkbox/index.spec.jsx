import React from 'react';
import renderer from 'react-test-renderer';

import Checkbox from '.';

describe('Checkbox', () => {
  it('renders correctly', () => {
    const input = {
      value: 'Value',
      onChange: () => {}
    };

    const meta = {
      error: '',
      touched: false
    };

    const tree = renderer.create(<Checkbox input={input} meta={meta} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
