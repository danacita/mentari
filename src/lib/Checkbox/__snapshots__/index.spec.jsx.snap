// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
<div
  className="sc-bdVaJa bSrzPW"
>
  <input
    className="sc-bwzfXH iesDFk"
    data-testid=""
    disabled={false}
    id={null}
    name={null}
    onChange={[Function]}
    type="checkbox"
    value="Value"
  />
  <label
    htmlFor={null}
  >
    <div
      className="sc-bxivhb iIpWIv"
    >
      
    </div>
  </label>
</div>
`;
